### ISSUE TITLE
Prefix autocompletion

### ISSUE THREAD

**Issue by ljos** [ljos/sparql-mode#30]
We should be able to add prefix autocompletion to the emacs mode. As YASGUI, we can use [prefix.cc](http://prefix.cc/) to complete the prefixes.

Look into:
- [prefix.cc](http://prefix.cc/)
- [company-mode](https://company-mode.github.io/)
- [autocomplete-mode](https://github.com/auto-complete/auto-complete)

# Acceptance criteria:
- [x] Complete PREFIX `namespace` with the full PREFIX definition of that namespace.
- [x] The user should be able to define their own prefixes.
- [x] Use prefix.cc optionally to complete the namespaces

## Optional:
- [x] Add company-mode plugin
- [ ] Add autocomplete plugin
- [ ] Cache prefix.cc prefixes

**Comment by ljos** [comment]
I use company-mode to provide the completion. I might do something with auto-complete-mode as well, but I think that might be better for someone actually using auto-complete-mode.

Implemented in fa4d20474a289d27131f5f34ee4993349fdcf113.

**Comment by VladimirAlexiev** [comment]
This code fetches prefix.cc every time:
```elisp
                 (url-insert
                   (url-retrieve-synchronously
                    "http://prefix.cc/popular/all.file.sparql")))
```
As an alternative, there is https://github.com/simenheg/rdf-prefix that uses a cached version (no internet needed), and I've recently updated it from the site.

Rather than Company, I use a function like this, which inserts proper syntax in sparql, turtle, shacl, shaclc; org and markdown code blocks:
```elisp
(defun my-rdf-prefix-insert (prefix)
  "Insert URI associated with PREFIX, using Turtle or SPARQL syntax, according to major mode or org-babel language.
Prefixes are defined by `rdf-prefix-alist' (prefix.cc) and complemented with `my-rdf-prefix-alist-wikidata'."
  (interactive
   (progn
     (eval-and-compile (require 'rdf-prefix))
     (unless (assoc "p" rdf-prefix-alist)
       (setq rdf-prefix-alist (append my-rdf-prefix-alist-wikidata rdf-prefix-alist)))
     (list (completing-read
            "RDF prefix: " rdf-prefix-alist nil t nil 'rdf-prefix-history))))
  (let* ((lang (and (eq major-mode 'org-mode)
                    ;; org-mode can tell us the language per code block
                    (nth 0 (org-babel-get-src-block-info 'light))))
         (shaclc (or (eq major-mode 'SHACLC-mode)
                     ;; (string-match-p "\\.shaclc$" (buffer-file-name))
                     (member lang '("shaclc" "SHACLC"))))
         (turtle (and (not shaclc)
                      (or (memq major-mode '(n3-mode Turtle-mode))
                          (member lang '("n3" "N3" "ttl" "turtle" "Turtle")))))
         (omn (or (eq major-mode 'omn-mode)
                  (member lang '("omn" "manchester")))) ; Manchester notation
          ;; markdown doesn't report the lang of ``` blocks, so I'll assume it's SPARQL
         (sparql (or (memq major-mode '(sparql-mode SPARQL-mode xsparql-mode markdown-mode))
                     (member lang '("sparql" "SPARQL")))))
    (insert (if turtle "@" "")
            (if shaclc "PREFIX " (if (or turtle sparql) "prefix " (if omn "Prefix: ") ""))
            prefix ": "
            (if (or shaclc turtle omn sparql) "<" "")
            (rdf-prefix-lookup prefix)
            (if (or shaclc turtle omn sparql) ">" "")
            (if turtle " ." ""))))
```
